app/notification.go

```diff
@@ -64,67 +59,57 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
 	}
 	channelMemberNotifyPropsMap := result.Data.(map[string]model.StringMap)
 
-	mentionedUserIds := make(map[string]bool)
-	threadMentionedUserIds := make(map[string]string)
+	mentions := &ExplicitMentions{}
 	allActivityPushUserIds := []string{}
-	hereNotification := false
-	channelNotification := false
-	allNotification := false
-	updateMentionChans := []chan *model.AppError{}
 
 	if channel.Type == model.CHANNEL_DIRECT {
 		otherUserId := channel.GetOtherUserIdForDM(post.UserId)
 
 		_, ok := profileMap[otherUserId]
 		if ok {
-			mentionedUserIds[otherUserId] = true
+			mentions.addMention(otherUserId, DMMention)
 		}
 
 		if post.Props["from_webhook"] == "true" {
-			mentionedUserIds[post.UserId] = true
+			mentions.addMention(post.UserId, DMMention)
 		}
 	} else {
-		keywords := a.getMentionKeywordsInChannel(profileMap, post.Type != model.POST_HEADER_CHANGE && post.Type != model.POST_PURPOSE_CHANGE, channelMemberNotifyPropsMap)
+		allowChannelMentions := a.allowChannelMentions(post, len(profileMap))
+		keywords := a.getMentionKeywordsInChannel(profileMap, allowChannelMentions, channelMemberNotifyPropsMap)
 
-		m := getExplicitMentions(post, keywords)
+		mentions = getExplicitMentions(post, keywords)
 
 		// Add an implicit mention when a user is added to a channel
 		// even if the user has set 'username mentions' to false in account settings.
 		if post.Type == model.POST_ADD_TO_CHANNEL {
-			val := post.Props[model.POST_PROPS_ADDED_USER_ID]
-			if val != nil {
-				uid := val.(string)
-				m.MentionedUserIds[uid] = true
+			addedUserId, ok := post.Props[model.POST_PROPS_ADDED_USER_ID].(string)
+			if ok {
+				mentions.addMention(addedUserId, KeywordMention)
 			}
 		}
 
-		mentionedUserIds, hereNotification, channelNotification, allNotification = m.MentionedUserIds, m.HereMentioned, m.ChannelMentioned, m.AllMentioned
-
 		// get users that have comment thread mentions enabled
 		if len(post.RootId) > 0 && parentPostList != nil {
 			for _, threadPost := range parentPostList.Posts {
```

Comment:
I'd do this in two parts, one just going through all the posts to get the users to notify, and then just iterate over that list. on a big thread this will save iterating over and over again for the same users.

we don't need to do that here (it probably is out of scope), but if you feel it makes sense we can create a new ticket for that.